### ISSUE TITLE
TWRP 2.8.7.0 Oneplus One fails to decrypt cm12.1 encryption

### ISSUE THREAD

**Issue by petto** [TeamWin/Team-Win-Recovery-Project#454]
Hi,

I am using TWRP 2.8.7.0 on a oneplus one with cm12.1 Nightly 2015-07-02.
Starting with this build I encrypted my device.
When I boot up TWRP, I am asked for my encryption password.
If I enter a wrong password, TWRP shows me that the password was wrong.
If I enter the correct password, TWRP proceeds but fails, unable to mount the data partition.

Trying to mount data with the mount screen doesn't work either.
If I try to check the checkbox on data the log gets filled with "unable to mount /data

Am I doing something wrong or is this a bug?
It seems that at least one other user had this bug on 2.8.6.1
http://forum.xda-developers.com/showpost.php?p=61497223&postcount=1134

Please see my recovery log if this helps,
Thanks!

__bionic_open_tzdata: couldn't find any tzdata when looking for localtime!
__bionic_open_tzdata: couldn't find any tzdata when looking for GMT!
__bionic_open_tzdata: couldn't find any tzdata when looking for posixrules!
Starting TWRP 2.8.7.0 on Sun Aug 30 08:55:43 1970
 (pid 195)
BOARD_HAS_NO_REAL_SDCARD := true
RECOVERY_SDCARD_ON_DATA := true
TW_NO_USB_STORAGE := true
TW_INCLUDE_CRYPTO := true
I:Found brightness file at '/sys/class/leds/lcd-backlight/brightness'
I:TWFunc::Set_Brightness: Setting brightness control to 255
Starting the UI...Pixel format: 1080x1920 @ 32bpp
Pixel format: RGB_565
Using qcomm overlay
Left Split=0
Right Split=0
Using qualcomm overlay
framebuffer: fd 3 (1080 x 1920)
=> Linking mtab
Renaming regular /etc/recovery.fstab -> /etc/recovery.fstab.bak
Moving /etc/twrp.fstab -> /etc/recovery.fstab
=> Processing recovery.fstab
I:Processing '/sbl1'
I:Processing '/rpm'
I:Processing '/tz'
I:Processing '/modem_st1'
I:Processing '/modem_st2'
I:Processing '/static_nv_bk'
I:Processing '/oppodycnvbk'
I:Processing '/aboot'
I:Processing '/boot'
I:Processing '/system'
I:Processing '/data'
I:Setting up '/data' as data/media emulated storage.
I:Created '/sdcard' folder.
I:Can't probe device /dev/block/mmcblk0p28
I:Unable to mount '/data'
I:Actual block device: '/dev/block/mmcblk0p28', current file system: 'ext4'
I:Can't probe device /dev/block/mmcblk0p28
I:Unable to mount '/data'
I:Actual block device: '/dev/block/mmcblk0p28', current file system: 'ext4'
get_crypt_ftr_info crypto key location: '/dev/block/mmcblk0p27'
I:Processing '/cache'
I:Processing '/misc'
I:Processing '/recovery'
I:Processing '/logo'
I:Processing '/reserve4'
I:Processing '/usb_otg'
I:Created '/usb_otg' folder.
I:Backup folder set to '/data/media/TWRP/BACKUPS/xxxxxxxx'
I:Settings storage is '/data/media'
Updating partition details...
I:Unable to mount '/usb_otg'
I:Actual block device: '', current file system: 'vfat'
...done
E:Unable to mount storage.
I:Unmounting main partitions...

Partition Logs:
/sbl1 | /dev/block/mmcblk0p2 | Size: 0MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p2
   Display_Name: sbl1
   Storage_Name: sbl1
   Backup_Path: /sbl1
   Backup_Name: sbl1
   Backup_Display_Name: sbl1
   Storage_Path: /sbl1
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/rpm | /dev/block/mmcblk0p6 | Size: 0MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p6
   Display_Name: rpm
   Storage_Name: rpm
   Backup_Path: /rpm
   Backup_Name: rpm
   Backup_Display_Name: rpm
   Storage_Path: /rpm
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/tz | /dev/block/mmcblk0p8 | Size: 0MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p8
   Display_Name: tz
   Storage_Name: tz
   Backup_Path: /tz
   Backup_Name: tz
   Backup_Display_Name: tz
   Storage_Path: /tz
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/modem_st1 | /dev/block/mmcblk0p10 | Size: 1MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p10
   Display_Name: modem_st1
   Storage_Name: modem_st1
   Backup_Path: /modem_st1
   Backup_Name: modem_st1
   Backup_Display_Name: modem_st1
   Storage_Path: /modem_st1
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/modem_st2 | /dev/block/mmcblk0p11 | Size: 1MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p11
   Display_Name: modem_st2
   Storage_Name: modem_st2
   Backup_Path: /modem_st2
   Backup_Name: modem_st2
   Backup_Display_Name: modem_st2
   Storage_Path: /modem_st2
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/static_nv_bk | /dev/block/mmcblk0p13 | Size: 10MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p13
   Display_Name: static_nv_bk
   Storage_Name: static_nv_bk
   Backup_Path: /static_nv_bk
   Backup_Name: static_nv_bk
   Backup_Display_Name: static_nv_bk
   Storage_Path: /static_nv_bk
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/oppodycnvbk | /dev/block/mmcblk0p12 | Size: 10MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p12
   Display_Name: oppodycnvbk
   Storage_Name: oppodycnvbk
   Backup_Path: /oppodycnvbk
   Backup_Name: oppodycnvbk
   Backup_Display_Name: oppodycnvbk
   Storage_Path: /oppodycnvbk
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/aboot | /dev/block/mmcblk0p5 | Size: 1MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p5
   Display_Name: aboot
   Storage_Name: aboot
   Backup_Path: /aboot
   Backup_Name: aboot
   Backup_Display_Name: aboot
   Storage_Path: /aboot
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/boot | /dev/block/mmcblk0p7 | Size: 16MB
   Flags: Can_Be_Backed_Up IsPresent Can_Flash_Img 
   Primary_Block_Device: /dev/block/mmcblk0p7
   Display_Name: Boot
   Storage_Name: boot
   Backup_Path: /boot
   Backup_Name: boot
   Backup_Display_Name: Boot
   Storage_Path: /boot
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/system | /dev/block/mmcblk0p14 | Size: 1303MB Used: 971MB Free: 331MB Backup Size: 971MB
   Flags: Can_Be_Mounted Can_Be_Wiped Can_Be_Backed_Up Wipe_Available_in_GUI IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p14
   Display_Name: System
   Storage_Name: System
   Backup_Path: /system
   Backup_Name: system
   Backup_Display_Name: System
   Storage_Path: /system
   Current_File_System: ext4
   Fstab_File_System: ext4
   Backup_Method: files

/data | /dev/block/mmcblk0p28 | Size: 57139MB
   Flags: Can_Be_Wiped Can_Be_Backed_Up Wipe_During_Factory_Reset Wipe_Available_in_GUI IsPresent Can_Be_Encrypted Is_Encrypted Has_Data_Media Can_Encrypt_Backup Use_Userdata_Encryption Is_Storage Is_Settings_Storage 
   Symlink_Path: /data/media
   Symlink_Mount_Point: /sdcard
   Primary_Block_Device: /dev/block/mmcblk0p28
   Crypto_Key_Location: /dev/block/mmcblk0p27
   Display_Name: data
   Storage_Name: Internal Storage
   Backup_Path: /data
   Backup_Name: data
   Backup_Display_Name: Data
   Storage_Path: /data/media
   Current_File_System: emmc
   Fstab_File_System: ext4
   Backup_Method: dd
   MTP_Storage_ID: 65537

/cache | /dev/block/mmcblk0p16 | Size: 503MB Used: 8MB Free: 495MB Backup Size: 8MB
   Flags: Can_Be_Mounted Can_Be_Wiped Can_Be_Backed_Up Wipe_During_Factory_Reset Wipe_Available_in_GUI IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p16
   Display_Name: Cache
   Storage_Name: Cache
   Backup_Path: /cache
   Backup_Name: cache
   Backup_Display_Name: Cache
   Storage_Path: /cache
   Current_File_System: ext4
   Fstab_File_System: ext4
   Backup_Method: files

/misc | /dev/block/mmcblk0p21 | Size: 1MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p21
   Display_Name: misc
   Storage_Name: misc
   Backup_Path: /misc
   Backup_Name: misc
   Backup_Display_Name: misc
   Storage_Path: /misc
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/recovery | /dev/block/mmcblk0p17 | Size: 16MB
   Flags: IsPresent Can_Flash_Img 
   Primary_Block_Device: /dev/block/mmcblk0p17
   Display_Name: Recovery
   Storage_Name: recovery
   Backup_Path: /recovery
   Backup_Name: recovery
   Backup_Display_Name: Recovery
   Storage_Path: /recovery
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/logo | /dev/block/mmcblk0p22 | Size: 16MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p22
   Display_Name: logo
   Storage_Name: logo
   Backup_Path: /logo
   Backup_Name: logo
   Backup_Display_Name: logo
   Storage_Path: /logo
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/reserve4 | /dev/block/mmcblk0p27 | Size: 64MB
   Flags: IsPresent 
   Primary_Block_Device: /dev/block/mmcblk0p27
   Display_Name: reserve4
   Storage_Name: reserve4
   Backup_Path: /reserve4
   Backup_Name: reserve4
   Backup_Display_Name: reserve4
   Storage_Path: /reserve4
   Current_File_System: emmc
   Fstab_File_System: emmc
   Backup_Method: dd

/usb_otg |  | Size: 0MB Used: 0MB Free: 0MB Backup Size: 0MB
   Flags: Can_Be_Mounted Can_Be_Wiped Wipe_Available_in_GUI Removable Is_Storage 
   Primary_Block_Device: /dev/block/sda1
   Alternate_Block_Device: /dev/block/sda
   Display_Name: USB-OTG
   Storage_Name: USB-OTG
   Backup_Path: /usb_otg
   Backup_Name: usb_otg
   Backup_Display_Name: USB-OTG
   Storage_Path: /usb_otg
   Current_File_System: vfat
   Fstab_File_System: vfat
   Backup_Method: files
   MTP_Storage_ID: 65538

I:Loading package: TWRP (/twres/ui.xml)
I:Load XML directly
I:Checking resolution...
I:Loading resources...
open_png /twres/images/twrplogo.png
libpng warning: iCCP: Not recognizing known sRGB profile that has been edited
open_png /twres/images/menu-button.png
open_png /twres/images/file.png
open_png /twres/images/folder.png
open_png /twres/images/slideout.png
open_png /twres/images/indeterminate001.png
open_png /twres/images/indeterminate002.png
open_png /twres/images/indeterminate003.png
open_png /twres/images/indeterminate004.png
open_png /twres/images/indeterminate005.png
open_png /twres/images/indeterminate006.png
open_png /twres/images/indeterminate007.png
open_png /twres/images/progress_empty.png
open_png /twres/images/progress_fill.png
open_png /twres/images/checkbox_empty.png
open_png /twres/images/checkbox_checked.png
open_png /twres/images/radio_empty.png
open_png /twres/images/radio_selected.png
open_png /twres/images/medium-button.png
open_png /twres/images/sort-button.png
open_png /twres/images/minus-button.png
open_png /twres/images/plus-button.png
open_png /twres/images/home-icon.png
open_png /twres/images/back-icon.png
open_png /twres/images/slider.png
open_png /twres/images/slider-used.png
open_png /twres/images/slider-touch.png
open_png /twres/images/unlock.png
open_png /twres/images/keyboard1.png
open_png /twres/images/keyboard2.png
open_png /twres/images/keyboard3.png
open_png /twres/images/keyboard4.png
open_png /twres/images/cursor.png
I:Loading variables...
I:Loading mouse cursor...
I:Loading pages...
I:PageSet::CheckInclude loading filename: '/twres/portrait.xml'
I:Loading included resources...
I:Loading included variables...
I:Loading mouse cursor...
I:Loading included pages...
I:Loading page main
I:Loading page main2
I:Loading page install
I:Loading page flash_confirm
I:Loading page flash_zip
I:Loading page flash_done
I:Loading page install_image
I:Loading page flashimage_confirm
I:Loading page clear_vars
I:Loading page confirm_action
I:Loading page action_page
I:Loading page singleaction_page
I:Loading page action_complete
I:Loading page filecheck
I:Loading page rebootcheck
I:Loading page wipe
I:Loading page advancedwipe
I:Loading page formatdata
I:Loading page formatdata_confirm
I:Loading page checkpartitionlist
I:Loading page partitionoptions
I:Loading page refreshfilesystem
I:Loading page selectfilesystem
I:Loading page backup
I:Loading page backupname1
I:Loading page backupname2
I:Loading page backupencryption
I:Loading page backupencryption2
I:Loading page checkbackuppassword
I:Loading page backup_run
I:Loading page restore
I:Unable to open '/data/media/TWRP/BACKUPS/xxxxxxxx'
I:Loading page restore_read
I:Loading page restore_decrypt
I:Loading page try_restore_decrypt
I:Loading page restore_select
I:Loading page renamebackup
I:Loading page restore_run
I:Loading page selectstorage
I:Loading page mount
I:Loading page usb_mount
I:Loading page usb_umount
I:Loading page system_readonly_check
I:Loading page reboot
I:Loading page settings
I:Loading page timezone
I:Loading page screen
I:Loading page Vibrate
I:Loading page advanced
I:Loading page partsdcard
I:Loading page htcdumlock
I:Loading page lock
I:Loading page filemanagerlist
I:Loading page filemanageroptions
I:Loading page choosedestinationfolder
I:Loading page filemanagerrenamefile
I:Loading page filemanagerrenamefolder
I:Loading page filemanagerchmod
I:Loading page filemanagerconfirm
I:Loading page filemanageracction
I:Loading page decrypt
I:Loading page decrypt_pattern
I:Loading page trydecrypt
I:Loading page terminalfolder
I:Loading page terminalcommand
I:Loading page sideload
I:Loading page fixperms
I:Loading page installsu
I:Loading page system_readonly
I:Switching packages (TWRP)
SELinux contexts loaded from /file_contexts
Full SELinux support is present.
Startup Commands: 
av.offload.enable=true
av.streaming.offload.enable=true
mm.enable.qcom_parser=3310129
mm.enable.smoothstreaming=true
ro.cm.device=bacon
ro.cm.display.version=12-20150620-UNOFFICIAL-bacon
ro.cm.version=12-20150620-UNOFFICIAL-bacon
ro.cm.releasetype=UNOFFICIAL
ro.qc.sdk.audio.fluencetype=fluence
ro.qc.sdk.camera.facialproc=true
ro.qc.sdk.gestures.camera=false
ro.sf.lcd_density=480
ro.com.google.clientidbase=android-oneplus
ro.com.android.dateformat=MM-dd-yyyy
ro.com.android.dataroaming=false
ro.com.android.wifi-watchlist=GoogleGuest
ro.gps.agps_provider=1
ro.qti.sdk.sensors.gestures=true
ro.qti.sensors.bte=true
ro.qti.sensors.vmd=true
ro.qti.sensors.gtap=true
ro.qti.sensors.ir_proximity=true
ro.url.legal=http://www.google.com/intl/%s/mobile/android/basic/phone-legal.html
ro.url.legal.android_privacy=http://www.google.com/intl/%s/mobile/android/basic/privacy.html
ro.boot.ftm=0
ro.boot.emmc=true
ro.boot.mode=recovery
ro.boot.selinux=permissive
ro.boot.baseband=msm
ro.boot.hardware=bacon
ro.boot.serialno=xxxxxxxx
ro.boot.bootdevice=msm_sdcc.1
ro.boot.startupmode=pwrkey
ro.data.large_tcp_window_size=true
ro.hwui.path_cache_size=32
ro.hwui.layer_cache_size=48
ro.hwui.texture_cache_size=72
ro.hwui.gradient_cache_size=1
ro.hwui.r_buffer_cache_size=8
ro.hwui.drop_shadow_cache_size=6
ro.hwui.text_large_cache_width=2048
ro.hwui.text_small_cache_width=1024
ro.hwui.text_large_cache_height=1024
ro.hwui.text_small_cache_height=1024
ro.hwui.texture_cache_flushrate=0.4
ro.oppo.rf_version=66
ro.oppo.pcb_version=12
ro.twrp.boot=1
ro.twrp.version=2.8.7.0
ro.wifi.channels=
ro.allow.mock.location=1
ro.board.platform=msm8974
ro.build.id=LRX22G
ro.build.date=Sat Jun 20 19:38:37 EDT 2015
ro.build.date.utc=1434843517
ro.build.host=build
ro.build.tags=test-keys
ro.build.type=eng
ro.build.user=jenkins
ro.build.display.id=bacon-eng 5.0.2 LRX22G ad739bd9ce test-keys
ro.build.product=A0001
ro.build.selinux=1
ro.build.version.sdk=21
ro.build.version.release=5.0.2
ro.build.version.codename=REL
ro.build.version.incremental=ad739bd9ce
ro.build.version.all_codenames=REL
ro.build.description=bacon-user 4.4.2 KVT49L XNPH25R release-keys
ro.build.fingerprint=oneplus/bacon/A0001:4.4.2/KVT49L/XNPH25R:user/release-keys
ro.build.characteristics=nosdcard
ro.config.ringtone=Orion.ogg
ro.config.alarm_alert=Hassium.ogg
ro.config.notification_sound=Argon.ogg
ro.crypto.state=encrypted
ro.dalvik.vm.native.bridge=0
ro.kernel.android.checkjni=1
ro.secure=0
ro.vendor.extension_library=libqti-perfd-client.so
ro.zygote=zygote32
ro.carrier=unknown
ro.cmlegal.url=https://www.cyanogenmod.org/docs/privacy
ro.product.cpu.abi=armeabi-v7a
ro.product.cpu.abi2=armeabi
ro.product.cpu.abilist=armeabi-v7a,armeabi
ro.product.cpu.abilist32=armeabi-v7a,armeabi
ro.product.cpu.abilist64=
ro.product.name=bacon
ro.product.board=MSM8974
ro.product.brand=oneplus
ro.product.model=A0001
ro.product.device=A0001
ro.product.locale.region=US
ro.product.locale.language=en
ro.product.manufacturer=OnePlus
ro.baseband=msm
ro.bootmode=recovery
ro.hardware=bacon
ro.opengles.version=196608
ro.qualcomm.bt.hci_transport=smd
ro.qualcomm.cabl=0
ro.qualcomm.perf.cores_online=2
ro.revision=0
ro.serialno=xxxxxxxx
ro.telephony.call_ring.multiple=0
ro.telephony.default_network=9
ro.bootloader=unknown
ro.debuggable=1
ro.modversion=12-20150620-UNOFFICIAL-bacon
ro.rommanager.developerid=cyanogenmod
ro.factorytest=0
ro.setupwizard.enterprise_mode=1
ro.use_data_netmgrd=true
net.bt.name=Android
net.change=net.bt.name
rol.ril.ext.ecclist=112,911,999,110,122,119,120,000,08,118
sys.keymaster.loaded=true
sys.listeners.registered=true
use.voice.path.for.pcm.voip=true
init.svc.adbd=running
init.svc.healthd=running
init.svc.ueventd=running
init.svc.qseecomd=running
init.svc.recovery=running
init.svc.set_permissive=stopped
rild.libpath=/system/vendor/lib/libril-qc-qmi-1.so
twrp.crash_counter=0
vold.decrypt=trigger_default_encryption
wifi.interface=wlan0
wifi.supplicant_scan_interval=15
audio.offload.pcm.16bit.enable=true
audio.offload.pcm.24bit.enable=true
audio.offload.buffer.size.kb=32
audio.offload.gapless.enabled=true
audio.offload.multiple.enabled=false
media.aac_51_output_enabled=true
dalvik.vm.isa.arm.features=div
dalvik.vm.heapsize=512m
dalvik.vm.dex2oat-Xms=64m
dalvik.vm.dex2oat-Xmx=512m
dalvik.vm.heapmaxfree=8m
dalvik.vm.heapminfree=2m
dalvik.vm.heapstartsize=16m
dalvik.vm.dex2oat-filter=interpret-only
dalvik.vm.heapgrowthlimit=192m
dalvik.vm.stack-trace-file=/data/anr/traces.txt
dalvik.vm.image-dex2oat-Xms=64m
dalvik.vm.image-dex2oat-Xmx=64m
dalvik.vm.image-dex2oat-filter=verify-none
dalvik.vm.heaptargetutilization=0.75
tunnel.audio.encode=true
camera2.portability.force_api=1
persist.gps.qc_nlp_in_use=0
persist.hwc.mdpcomp.enable=true
persist.sys.dun.override=0
persist.sys.usb.config=mtp
persist.sys.dalvik.vm.lib.2=libart.so
persist.sys.root_access=0
persist.sys.isUsbOtgEnabled=true
persist.sys.recovery_update=false
persist.data.netmgrd.qos.enable=true
persist.data.tcpackprio.enable=true
persist.audio.fluence.voicecall=true
persist.radio.add_power_save=1
persist.radio.apm_sim_not_pwdn=1
persist.timed.enable=true
service.adb.root=1
keyguard.no_require_sim=true
recovery.perf.mode=0
telephony.lteOnGsmDevice=1

I:Copying file /cache/recovery/log to /cache/recovery/last_log
I:Is encrypted, do decrypt page first
I:Switching packages (TWRP)
I:Set page: 'decrypt'
I:Set page: 'trydecrypt'
I:operation_start: 'Decrypt'
CONFIG_HW_DISK_ENCRYPTION present
hexadj_passwd converting to lower case for hardware disk crypto.
has_hw_crypto is 1
crypt_ftr->fs_size = 117021080
Using scrypt with keymaster for cryptfs KDF
Signing safely-padded object
create_crypto_blk_dev extra_params set to fde_enabled
load_crypto_mapping_table using req-crypt
load_crypto_mapping_table: target_type = req-crypt
load_crypto_mapping_table: real_blk_name = /dev/block/mmcblk0p28, extra_params = fde_enabled
Password matches
Data successfully decrypted, new block device: '/dev/block/dm-0'
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
Updating partition details...
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Unable to mount '/usb_otg'
I:Actual block device: '', current file system: 'vfat'
...done
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
E:Unable to mount storage.
I:Unmounting main partitions...
tw_get_context failed to get selinux context
I:Failed to get default contexts and file mode for storage files.
I:Set page: 'main'
I:Set page: 'clear_vars'
I:Set page: 'main2'
I:operation_end - status=0
I:Set page: 'main'
I:Set page: 'clear_vars'
I:Set page: 'main2'
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
E:Unable to mount settings storage during GUI startup.
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
E:Unable to mount /data/media/TWRP/.twrps when trying to read settings file.
I:Attempt to load settings from settings file...
I:Settings file '/data/media/TWRP/.twrps' not found.
I:Copying file /etc/recovery.fstab to /cache/recovery/recovery.fstab
I:Version number saved to '/cache/recovery/.version'
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Unable to mount '/usb_otg'
I:Actual block device: '', current file system: 'vfat'
__bionic_open_tzdata: couldn't find any tzdata when looking for CST6CDT,M3.2.0,M11.1.0!
__bionic_open_tzdata: couldn't find any tzdata when looking for posixrules!
I:TWFunc::Set_Brightness: Setting brightness control to 255
I:TWFunc::Fixup_Time: Pre-fix date and time: 1970-08-30--03-56-44
I:TWFunc::Fixup_Time: Setting time offset from file /sys/class/rtc/rtc0/since_epoch
I:TWFunc::Fixup_Time: will attempt to use the ats files now.
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Enabling MTP during startup
I:Starting MTP
Starting MTP
I:Can't probe device /dev/block/dm-0
E:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Can't probe device /dev/block/dm-0
I:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
MTP Enabled
I:Check_Lifetime_Writes result: '656020'
I:Can't probe device /dev/block/dm-0
E:Unable to mount '/data'
I:Actual block device: '/dev/block/dm-0', current file system: 'ext4'
I:Switching packages (TWRP)
created new mtpserver object
MtpServer::run fd: 13
I:TWFunc::Set_Brightness: Setting brightness control to 5
I:TWFunc::Set_Brightness: Setting brightness control to 0
I:TWFunc::Set_Brightness: Setting brightness control to 255
I:Set overlay: ''
I:TWFunc::Set_Brightness: Setting brightness control to 5

**Comment by Maxqia** [comment]
Fails for me too

**Comment by xitation** [comment]
Failing for me on my Nexus 5 (hammerheadcaf) running CM12.1-20150620. I'm using pattern unlock.

**Comment by petto** [comment]
Can we expect any status update or some kind of "yes, it's a known issue that we are working on" ?

**Comment by Dees-Troy** [comment]
I'm not sure how you guys think this works. TWRP is free open source software. I'm pretty much the only TWRP developer that works on decrypting data partitions. I work on TWRP in my spare time when I feel like it. I guess you can say that I'm aware that decrypt doesn't work for some people on the OnePlus One since it's logged here in the issues log. I'm not currently working on it and I have no idea if or when I'll work on it again.

**Comment by petto** [comment]
Thanks for this update, I can totally understand this and for me thats okay :-)

I just wasn't sure that this issue got through to you.
Wouldn't be the first time a bug report has never been read. (not by you, no offense!)

I am very happy to have software like twrp thats developed in someones spare time.
Without this software, we woulnd't be able to do many things that we can do today.

If you need any more information or testing help when you are looking into it, please let me know, I'll help.

**Comment by optimumpr** [comment]
This is not an issue specific to Oneplus only.  I am yet to see one single device where TWRP can mount encrypted data partition in lollipop.  It did work fine on kitkat, but not lollipop...

**Comment by kendallgreen** [comment]
Dees-Troy, can you let us know when you get TWRP tol handle an encrypted CM12.1 oneplus phone?

**Comment by agherzan** [comment]
Reproduced on LG G4 (Android 6).

**Comment by davidtorcivia** [comment]
Reproduced on LG G4 H815 w/ CM12.1.

**Comment by kraemert** [comment]
Reproduced on Sony Xperia Z3 Compact on Android 5.1.1

**Comment by raaomoka** [comment]
Wow, this is still going on up till today? Guess I'll have to switch to different recovery that actually is being supported...

**Comment by nailyk-fr** [comment]
Hi, 

I own a xperia z3 and experiencing the same issue so I work to get uncrypt /data working. 
In fact, z3 and OPO are based on the same soc: msm8974 witch embed a hardware keystore. What I understand is the keystore is unlocked by our pin/pass/...:
http://nelenkov.blogspot.fr/2013/08/credential-storage-enhancements-android-43.html #The hardware in 'hardware-backed'.
After including keymaster blobs and keystore manager, twrp is able to uncrypt /data. I haven't access to a oneplus one but I can try to build a twrp or help someone to do this. 
As shown in my recovery.log: 
P.S.: my recovery.log http://paste.debian.net/356228/
I'm pretty sure cm12.1 encrypt over f2fs filesystem. 
For now my 'code' modifications are not public because it's q&d job and I need some help to make this better. 
I hope my modifications are also useful for OPO. Are someone interested to work with me and/or test with the oneplus one?

**Comment by raaomoka** [comment]
Hi,

I was experiencing this issue 'cause I was using TWRP v2.8.6.0 (after reading that later versions don't update the baseband and whatnot). I updated to v2.8.7.0 and it's working with the encryption...speaking of which, I think there should be an easy unencrypt button implemented in Cyanogen OS...gonna suggest that on Jira.

**Comment by nailyk-fr** [comment]
@raaomoka : 

> I think there should be an easy unencrypt button implemented in Cyanogen OS

From this: https://source.android.com/security/encryption/ I think it's not possible. 
Glad that twrp is now working on encrypted OPO with CM>12 / android > 5. 
I post here because the issue was still open. 
Sorry for inconvenience.

**Comment by raaomoka** [comment]
@nailyk-fr Well, according to the following, Samsung SmartPhones have already got such function:

http://m.androidcentral.com/how-disable-encryption-android-lollipop

**Comment by Micha-Btz** [comment]
@nailyk-fr can you please describe it more in detail what you did?
because my Sony Z1 Compact is also not able to decrypt my data partition.

edit: the z1c is also based on msm8974

**Comment by optimumpr** [comment]
I can help, many would, but you need to post your changes.

**Comment by nailyk-fr** [comment]
Hi, 
*_What Is done: *_
As xperia z3 is (for now :smile: ?) not supported by twrp, I start from a cyanogenbuild tree and add this manifest [1](http://paste.debian.net/359557/)
Diff for the 'device_sony_shinano-common' from the official cyanogen github (sorry because twrp should be build from omnirom) is here: [2](http://paste.debian.net/359538/)

*_How it is done: *_
In a global way, what I understand: our phones (based on msm8974 soc) have hardware keystore. Then, the passphrase unlock the keystore which contain the hashed default password for the dmcrypted data partition. 
When the system is started, all the binaries blobs are available so it should be the same for the recovery: 
- First try I made: compile twrp with
  - `TW_INCLUDE_CRYPTO := true`
  - `TARGET_HW_DISK_ENCRYPTION := true`
- (The last one seems to be useless)
- Then, I manually mount /system and twrp was able to decrypt the /data partition. 
- From this commit [3](https://github.com/TeamWin/android_device_oneplus_bacon/commit/52d6a76793170ca4c76de790a8841d87856253a8) a 'first' blob list [4](http://paste.debian.net/359561/) appears for the z3
- After copying all this blobs the 'keystore application' is also needed (see LOCAL_POST_INSTALL_CMD in the diff). 
- Finally, symlink all the created folders in the recovery ramdisk to the /system mount folder. I think symlink is needed in case of /system need to be mounted: there is no lock on the mount point and with this 'q&d' patch blobs are available with /system mounted or not. 

Thanks for reading, I think all is here. As i'm not at home I can't start from scratch to check the process but seems all my work is here. Hope it can be helpful for somebody and sorry for out-topic post. 
Maybe, if we can confirm that all 8974 soc decrypt /data the same way, we can open a 'global issue'? 

P.S.: Phones listed in this issue: 
-  xitation Nexus 5 | _MSM8974 Snapdragon 800_ | hardware keystore
-  agherzan: LG G4 & davidtorcivia  LG G4 H815 | _MSM8992 Snapdragon 808_ | can't find quickly
-  wallsofjericho Sony Xperia Z3 Compact | _MSM8974AC Snapdragon 801_ | hardware keystore
-  Micha-Btz Sony Z1 Compact | _MSM8974 Snapdragon 800_ | hardware keystore
-  nailyk Sony Xperia Z3 | _MSM8974AC Snapdragon 801_ | hardware keystore
-  obviously 1+1 | _MSM8974AC Snapdragon 801_ | hardware keystore

Maybe it is related to the snapdragon 800 serie? 

P.S.2: I need help because it's first time I compile with NDK and I'm not familiar with dot mk files. So I do lot of 'workarounds' to be able to copy blobs, create symlinks on a automated way [2](http://paste.debian.net/359538/) and think it can be done cleaner. (actually I build with 'mka recovery-zip' )

P.S.3: my diff [2](http://paste.debian.net/359538/) is polluted by 'auto-update-zip-recovery' addition, sorry. I try to make a cleaner patch asap. 

[1](http://paste.debian.net/359557/) http://paste.debian.net/359557/
[2](http://paste.debian.net/359538/) http://paste.debian.net/359538/
[3](https://github.com/TeamWin/android_device_oneplus_bacon/commit/52d6a76793170ca4c76de790a8841d87856253a8) [https://github.com/TeamWin/android_device_oneplus_bacon/commit/52d6a76793170ca4c76de790a8841d87856253a8](https://github.com/TeamWin/android_device_oneplus_bacon/commit/52d6a76793170ca4c76de790a8841d87856253a8)
[4](http://paste.debian.net/359561/) http://paste.debian.net/359561/

**Comment by optimumpr** [comment]
I don't think I follow you.  TWRP asks you for pin/password on boot and then fails to decrypt with the correct password.  Are you saying you entered 'cancel" at password prompt, then mounted system, then attempted to decrypt and it worked? I just checked on my device Xperia Z1 and it fails to decrypt. 

I have found the best way to build TWRP is in a rom build environment where you have everything  including kernel that you need.  I just get the latest twrp sources from omni or slimrom, do build.envelope, lunch and 'mka recoveryimage'.

Perhaps the culprit is in android_device_honami/shinano ot whatever.  Could you post entries in  your device recovery/root/init.recovery.----.rc file? Thanks.

**Comment by nailyk-fr** [comment]
@optimumpr 

> TWRP asks you for pin/password on boot and then fails to decrypt with the correct password. Are you saying you entered 'cancel" at password prompt, then mounted system, then attempted to decrypt and it worked? I just checked on my device Xperia Z1 and it fails to decrypt. 

Yes it's what I've done: 
1. Identify the dm-0 filesystem when cyanogenmod is launched (since CM12.0 seems to be f2fs): 
[mount | grep "dm-0"](https://cloud.githubusercontent.com/assets/16372405/12121033/7d178662-b3d3-11e5-9058-1f0605416145.png)
2. change my BoardConfig(Common).mk to include this settings: 
`TW_INCLUDE_CRYPTO := true`
`TARGET_HW_DISK_ENCRYPTION := true`
3. replace 
`TW_CRYPTO_FS_TYPE := "ext4"`
with 
`TW_CRYPTO_FS_TYPE := "f2fs"`
4. then 'mka recoveryimage -j#' and flash it. 

With this twrp and the good password (I use a PIN), the decryption doesn't work out-of-box. But after hitting 'Cancel' for the /data password, mounting /system, then click on /decrypt everything goes fine. 

@optimumpr can you post your recovery.log? 
Because I remember that was difficult to enable CONFIG_HW_DISK_ENCRYPTION wich is a twrp compilation flag needed to enable the keystore functionalities (if I successful read cryptfs.c) in twrp. Think it's enabled by adding TARGET_HW_DISK_ENCRYPTION in BoardConfig.mk. If it's ok, you should see this line: 
`CONFIG_HW_DISK_ENCRYPTION present` 
in your recovery.log. 

@optimumpr 

> I have found the best way to build TWRP is in a rom build environment where you have everything including kernel that you need. I just get the latest twrp sources from omni or slimrom, do build.envelope, lunch and 'mka recoveryimage'.

Really curious, I can't find CONFIG_HW_DISK_ENCRYPTION flag on recent twrp versions: [TeamWin source search](https://github.com/TeamWin/Team-Win-Recovery-Project/search?utf8=%E2%9C%93&q=CONFIG_HW_DISK_ENCRYPTION&type=Code) . After a little research I realize the source for twrp I use is [omnirom](https://github.com/omnirom/android_bootable_recovery/blob/android-5.1)
In this version [cryptfs.c](https://github.com/omnirom/android_bootable_recovery/blob/android-5.1/crypto/lollipop/cryptfs.c) include this flag. As you are using ominrom it should work :blush: 

@optimumpr 

> Could you post entries in your device recovery/root/init.recovery.----.rc file?

[init.recovery.qcom](http://paste.debian.net/360219/) 

@optimumpr 

> I don't think I follow you. [...] Thanks.

Glad (try) to help, but I'm sorry for misunderstanding. Actually I'm not comfortable with English language so I'm not able to give the right 'message' for now. I work to express myself less confusing :blush:

**Comment by nailyk-fr** [comment]
@optimumpr 
Hi, 
I build a 'experimental-testing-maybe_it_burn_your_phone' recovery for your device. 
Seems there is no hardware keystore on Honami. Apparently there is multi versions of msm8974. Maybe more informations [here](https://www.qualcomm.com/products/snapdragon/tools) & [snapdragon wikipedia](https://en.wikipedia.org/wiki/List_of_Qualcomm_Snapdragon_devices#Snapdragon_801)
Looking for a list but can't find it. 

This [FOTAKernel](http://s000.tinyupload.com/?file_id=57469159288064027993) is obviously for unlocked boot-loaders and is experimental as I haven't Z1(honami). It should work only after mounting /system and on f2fs dmcrypted partition, if it work :blush:

**Comment by optimumpr** [comment]
Thanks for your efforts. Unfortunately, I can't test a F2FS version. I do have all the appropriate flags.  I think the culprit is to have appropriate libs/blobs. I will try to add blobs entries in boardconfg and compile.  I think you also need CONFIG_HW_DISK_ENCRYPTION=y in kernel.

**Comment by nailyk-fr** [comment]
Hi
Think your right: need the appropriate libs. 
After a little research found this [blog](http://nelenkov.blogspot.fr/2013/08/credential-storage-enhancements-android-43.html?m=1). 
First screenshot is very interesting, it shown if you are using hardware or software keystore (Parameters, Security, Credentials, Storage type). He also says (in the 'The hardware in 'hardware-backed'' section) the hardware keystore is in the Trust-Zone (see /dev/block/platform/*.1/by-name/tz) and for qualcomm it's available through /dev/qseecom interfaced by the libQSEEComAPI.so . 
Think that a start. 
@optimumpr 

> I think you also need CONFIG_HW_DISK_ENCRYPTION=y in kernel.

I grep this flag in defconfigs but can't find it. Is there a way to check kernel compilation flags once phone is started (in recovery)? 
Thanks

**Comment by Micha-Btz** [comment]
i use https://github.com/erikcas/android_device_sony_rhine-common/blob/android-6.0-cas/BoardConfigOmni.mk which is reported to work with encryption in omni rom. There seems to be a special thing with cyanogenmod. i haven't tested omni rom so far but i will do it the next days.

for me it is not working also not with mount /system at first.

**Comment by nailyk-fr** [comment]
@Micha-Btz

> i use https://github.com/erikcas/android_device_sony_rhine-common/blob/android-6.0-cas/BoardConfigOmni.mk which is reported to work with encryption in omni rom. There seems to be a special thing with cyanogenmod. i haven't tested omni rom so far but i will do it the next days.

I think you are right. Cyanogen use f2fs filesystem over encryption, it seems other roms use ext4. 
The twrp posted on 6-jan is build with f2fs support. 
Would you like check your filesystem with [mount | grep "dm-0"](https://cloud.githubusercontent.com/assets/16372405/12121033/7d178662-b3d3-11e5-9058-1f0605416145.png) ? 

I remember you use a Z1c. Do you need a special build or is it compatible between Z1&Z1c ? 
[Experimental FOTAKernel twrp for **Z1** (honami)](http://s000.tinyupload.com/index.php?file_id=41486771446144427844) with ext4 over dmcrypt. No support for backlight. Blobs inculded. (MD5: ccc467547f0f4147343cf743e8a38034 )

Thanks for your feedback.

**Comment by Micha-Btz** [comment]
for me ext4 is correct. i don't know if z1/z1c are equal but i don't think so.
i use recovery for the sony real recovery partition not for the fota.
http://developer.sonymobile.com/2015/05/18/recovery-mode-now-available-for-a-range-of-unlocked-xperia-devices-video/

**Comment by optimumpr** [comment]
It does not matter whether you have a dedicated recovery partition or FOTA or your recovery is in boot image. The reason TWRP can't decrypt data is that Qseecom interface is not brought up, so that the master key is signed by a certificate that is sitting in hardware (that's why TWRP reports incorrect password, because a required element, hardware signing, is not present.  For that you either need changes in cryptfs.c (in TRWP) and/or in your device tree.  Changes in TWRP will never be made, because this may break other devices. 

Anyway, from a security point, it is not so bad that TWRP can't decrypt.  You can still upgrade your rom since System partition is not encrypted, and you can backup your data from a booted device (when data is plain) via adb.

**Comment by nailyk-fr** [comment]
@Micha-Btz 
Thanks for the link and the info. I update my bootloader and it implies some minor change to work. As expected recovery in FOTAKernel would work with new bootloader (v27 for s3) with the same flashing process (fastboot flash FOTAKernel recovery.img / fastboot flash recovery recovery.img ): only the phone boot process as been changed. 
I try to build a test recovery for your z1c asap. 
Regards.

**Comment by nailyk-fr** [comment]
@Micha-Btz 
Here is experimental 
  [twrp for z1c (amami)](http://s000.tinyupload.com/index.php?file_id=43077224187021190206) (md5: 00629a4cb8b4fec266e2d77a9f85d7fa)
  [twrp for z1 (honami)](http://s000.tinyupload.com/index.php?file_id=53014812833538768694) (md5: cac20de238837c5755aa3ba6466f8e23)
This should be compatible with old and new bootloader. 
- First I recommend to backup your actual recovery because I'm not really sure this one will boot. 
- It should be flashed by 'fastboot flash recovery FOTAKernel_amami.20160109.img'
- If it fail to decrypt at first time can you try to mount /system and try again?
- Can you provide full recovery.log, full dmesg please and the rom installed on your phone? 

Thanks in advance.

**Comment by Micha-Btz** [comment]
@optimumpr 
as far as you see here https://gerrit.omnirom.org/#/c/13673/3
i think the changes are made for twrp.
for me i want to have an recovery with encryption support because i want to test some other roms and so a rollback is not possible.
@nailyk-fr 
i will test your recovery and report back.

**Comment by Micha-Btz** [comment]
@nailyk-fr your recovery is working also encryption is working.
here the recovery.log https://gist.github.com/Micha-Btz/a72c51f1fffb1e48384c

But some thing makes me totally crazy my data partition is encrypted without asking me for a password.
The log says encrypted with default password, what the hell it is?
What is the default password?

**Comment by optimumpr** [comment]
@nailyk-fr 
Congratulations on having this working on Z1c. Unfortunately, the changes you posted so far do not result in this working on Z1.  Perhaps you have made some changes that are not posted yet? Could you please post the latest changes? Thanks.

@Micha-Btz 
If you don't have a pin setup before encrypting, then your data partition is encrypted with default password, which is literally 'default_password'. Depending on your rom, you might have settings (in security) to have the system ask for default password, but that is no security at all.  You should do encryption over adb to have one long password for boot/encryption and another short one to unlock screen.  You can head for XDA Security discussion thread where I have a detailed guide on that.

**Comment by nailyk-fr** [comment]
@Micha-Btz 
Glad it works, we are on the right way! 
I don't understand way you are not asked for your pin/password. As @optimumpr says, the dmcrypt is always encrypted with 'default_password'. This password is stored salted (by your pin/password) in the keystore. When you change your pin/password, the key is replaced in the keystore (but the 'real password for the partition' remain unchanged). This prevent the dm-crypted partition to be re-write on each pin/password change. [More info](https://source.android.com/security/encryption/index.html#how_android_encryption_works)
@optimumpr 
Did you try the z1 version a post earlier? 
I doesn't understand why you can't reproduce build. Think your right I miss something. When copying my workarounds to 'rhine-common' I clean up a little the recovery/Android.mk I make so it should be easier to read now. 

What is the easiest for you? Archive of device/sony/rhine-common? Patch for the different (all?) devices-tree? Do you need my manifests files?

**Comment by optimumpr** [comment]
@nailyk-fr 
I would prefer device/sony/rhine/common. I wonder if there were any differences for Z1 that are not on your paste.debian.net?

**Comment by nailyk-fr** [comment]
@optimumpr 

> I would prefer device/sony/rhine/common. I wonder if there were any differences for Z1 that are not on your paste.debian.net?

I double check and it should have no difference (except the auto-zip generation) :disappointed:  I cp my dirty recovery/Android.mk, cp the blobs, adapt blobs path for rhine and build. Really sorry...

z1: [android_device_sony_honami.patch](http://paste.debian.net/364556) / [original](https://github.com/CyanogenMod/android_device_sony_honami/tree/cm-12.1)
z1c: [android_device_sony_amami.patch](http://paste.debian.net/364567) / [original](https://github.com/CyanogenMod/android_device_sony_amami/tree/cm-12.1)
rhine-board: [android_device_sony_rhine-common.tar.gz](http://s000.tinyupload.com/index.php?file_id=33846689359976247800) / [original](https://github.com/CyanogenMod/android_device_sony_rhine-common/tree/cm-12.1)
No difference with [android_bootable_recovery](https://github.com/omnirom/android_bootable_recovery)

I don't know what other check can be done...
Hope I don't miss anything this time :confused:

**Comment by optimumpr** [comment]
OK. There were a few flags I was missing.  However, I have just compiled recovery according to your latest post and the result unfortunately is the same. I do have a long password (over 50 characters), but it works without a problem with the system.  I can't run adb logcat while in recovery.  Is there any other way to run it in recovery?

**Comment by optimumpr** [comment]
Here is dmesg which I ran from recovery. It does show Qsecom brought up, but then later it says can't find ext4 file system...

http://paste.debian.net/364605/

**Comment by nailyk-fr** [comment]
@optimumpr 

> I can't run adb logcat while in recovery. Is there any other way to run it in recovery? 

Did you try, in the recovery terminal, something like `adb logcat > /external_sd/log_decrypt` ?

> I do have a long password (over 50 characters)

I read somewhere twrp have a length limitation, I check the code asap. Is there any chance you test with a shorter password? (Think it is safer to backup your data first)

> It does show Qsecom brought up, but then later it says can't find ext4 file system...

Here is my [dmesg](http://paste.debian.net/364639/) & [recovery.log](http://paste.debian.net/364641/) for comparison. I use 8 digit pin. 
Did you check your rom dmcrypt header location? It is located on a flag in the fstab. Eg for [CM12.1](https://github.com/CyanogenMod/android_device_sony_rhine-common/blob/cm-12.1/rootdir/fstab.qcom): it is located on the userdata partition footer. But it can be locate somewhere else. 
Can you provide the recovery.log associated to the dmesg? 

Otherwise your dmesg is very clean and seems you use a self build gcc, nice job!

**Comment by optimumpr** [comment]
@nailyk-fr 
No problem with dmcrypt header: it is in footer. Is there a special dmesg command to run to get a recovery log? Would that be dmesg --userspace? 

I also have a few changes made in cryptfs.c and cryptfs.h, which results in a different vold binary, but I always incorporate the same changes into twrp corresponding files. This is why I can't test your build...

**Comment by nailyk-fr** [comment]
@optimumpr 
you can obtain the recovery.log on your sdcard by clicking [advanced](http://dl-1.va.us.xda-developers.com/3/1/3/6/6/7/5/Screenshot_2015-01-27-10-00-18.png?key=_tE0MhvtesjBoNH7KdVYFg&ts=1452547992) then [copy log to sd](http://dl-1.va.us.xda-developers.com/3/1/3/6/6/7/6/Screenshot_2015-01-27-10-00-25.png?key=uW3itbfUVCjD4Y2_rpJMYg&ts=1452548042), you get a recovery.log file if your external sdcard is mounted.

**Comment by optimumpr** [comment]
@nailyk-fr 
Because data is not mounted and Sdcard is emulated (it is in data), TWRP can't save it there. Where is the original recovery.log located? Anyway, I will look into it further.  Thanks.

**Comment by nailyk-fr** [comment]
I edit my post after, sorry. It copy the log on external sd. The log is temporary located on /tmp.

**Comment by optimumpr** [comment]
OK. Here is an excerpt from recovery.log:  
config_hw_disk_encryption present
password not correctly hex encoded
has_hw_crypto os 0
could not open keymaster device in keystore (operation not permitted)
failed to init keymaster
signing failed

That is strange because selinux is hard disabled in my kernel

**Comment by nailyk-fr** [comment]
@optimumpr 

> password not correctly hex encoded

I attempt to read the cryptfs.c/h and have a supposition: If your password length is bigger than the limit, the ascii --> hex conversion can't be done. cryptfs.c line [1282](https://github.com/omnirom/android_bootable_recovery/blob/android-5.1/crypto/lollipop/cryptfs.c#L1282) & cryptfs.h line [42](https://github.com/omnirom/android_bootable_recovery/blob/android-5.1/crypto/lollipop/cryptfs.h#L42) but really not sure of my reading :smile:

**Comment by optimumpr** [comment]
@nailyk-fr 
Here is the complete recovery log: http://paste.debian.net/364687/

My defined MAX_KEY_LEN is set to 96 in both /system/vold/, as well as in TWRP, so it shouldn't be a problem. What do you think the reason for unable to open keymaster (permission denied) is?

**Comment by nailyk-fr** [comment]
@optimumpr 
I'm pretty sure the `Password not correctly hex encoded.` is caused by the password length: when I type +50 character on the password I have the same error. If I type a shorter bad password I don't have. I read a lot of comment in the code that will show the password in clear. I will try to identify the function causing the trouble. 
For the permission denied I'm looking but I remember something about selinux and keymaster. I try to disable selinux on my build and give you feedback.

**Comment by optimumpr** [comment]
@nailyk-fr 
I have tried with a 8 character password and this time I got
I:operation_start: 'Decrypt'
CONFIG_HW_DISK_ENCRYPTION present
has_hw_crypto is 0
crypt_ftr->fs_size = 25178047
Using pbkdf2 for cryptfs KDFcreate_crypto_blk_dev extra_params set to empty string
get_dm_crypt_version flag: 0, name: 'verity'
get_dm_crypt_version flag: 1, name: 'crypt'
Enabling support for allow_discards in dmcrypt.
load_crypto_mapping_table using crypt
load_crypto_mapping_table: target_type = crypt
load_crypto_mapping_table: real_blk_name = /dev/block/mmcblk0p25, extra_params = 1 allow_discards
Error temp mounting decrypted block device '/dev/block/dm-0'

It  can't access keymaster for some reason...

P.S. Keep in mind that operation works flawlessly on straight boot (without TWRP)

**Comment by nailyk-fr** [comment]
from the part of log you post the dmcrypt was successfully created: 
Enabling support for allow_discards in dmcrypt.
&
load_crypto_mapping_table: real_blk_name = /dev/block/mmcblk0p25, extra_params = 1 allow_discards
but cannot create the /dev/block/dm-0 pseudo file? 
Flashing right now my version without selinux.

**Comment by nailyk-fr** [comment]
When hitting 'ok' on the decrypt form, twrp crash.

**Comment by optimumpr** [comment]
@nailyk-fr 
Interesting.  What I am missing is the "signing safely padded object" with keymaster
Whatever it is (selinux on or off), the problem is in TWRP, because everything works fine on a straight boot.

**Comment by optimumpr** [comment]
Well. If I don't succeed, that would mean my device is virtually impossible to break by using third party interfaces. LOL...

**Comment by nailyk-fr** [comment]
@optimumpr 

> Whatever it is (selinux on or off)

You are right, I revert and it crash the same way... 
Maybe juste a unix permission file problem? Or a flag in your kernel? Do you want the def_config I use?

**Comment by optimumpr** [comment]
@nailyk-fr 
Yes, post  your def_config, but I don't think it is kernel, because it wouldn't then work on straight boot. Now, I did 'make clean' and am trying to build the whole rom and now it tells me "No rule to make target `device/sony/rhine-common/rootdir/twrp.fstab', needed by`/home/Android/Tesla/out/target/product/honami/recovery/root/etc/twrp.fstab. Strange. I need to look at device tree again...

**Comment by nailyk-fr** [comment]
as I only build 'recoveryimage', in the recovery/Android.mk there is a workaround for the twrp.fstab copying. Maybe the source file is in the wrong place? I move it from device/sony/amami/rootdir to device/sony/rhine-common. 
Oo Did your report my fstab modifications or you use your old fstab in your $(OUT) tree? if the '-length=xxxxx' is missing the decryption can't work ;)

**Comment by optimumpr** [comment]
I have twrp.fstab in both /device/sony/honami and /device/rhine-common.  I think keymaster is not being built correctly. You put there libs from your Z3 rom. I will try to put the same libs from my rom and see how it goes.

**Comment by nailyk-fr** [comment]
@optimumpr 
in the device tree I give you that is rhine blobs. I replace all firmware but a double check is always a good idea :smile: . And seems it works for the z1c. Good luck for the whole build. 
Here is a [log](http://paste.debian.net/364741/) with plain (bad) password. Seems they are not truncated. 
The [defconfig](http://paste.debian.net/364752/) (not modified so it should be the same as [CM](https://github.com/CyanogenMod/android_kernel_sony_msm8974/blob/cm-12.1/arch/arm/configs/cm_rhine_amami_row_defconfig) ). 
Hope this help.

**Comment by optimumpr** [comment]
Thanks. I am missing 'Using scrypt with keymaster for cryptfs KDF ".  In my TWRP it uses scrypt with pkdf2.  I'll keep trying. Good night...

**Comment by Micha-Btz** [comment]
why you don't fork the sony rhine repo and insert your patches there and push it back to github so we all can take a look at your changes. would be the best method i think.

**Comment by nailyk-fr** [comment]
@optimumpr 
Have you try to decrypt with `twrp decrypt <password>` cause if you encrypt with `vdc cryptfs encrypt <passwd>` I suppose your password in [not hex encoded](http://android.stackexchange.com/questions/117405/how-do-you-encrypt-your-device-running-cyanogenmod-12-1/119455#119455). Maybe the twrp command line could be a workaround. 
@Micha-Btz 
Because I'm really not comfortable with the git command and I'm always afraid to push on a repository I don't own. But I'm here to learn so I will do that :smile: 
Have you try to build twrp with the patch I post earlier?

**Comment by optimumpr** [comment]
@nailyk-fr 

I got your github notification. Thanks. It is not working for me and the logs are inconsistent. I have tried to disable selinux in TWRP; then I chmod /vendor/keymaster/ in recovery to 755 and it still is not working. One time I got 'can't find keymaster module'; the other time I had invalid hex string failed to convert passwd from hex. That did not make sense, because password is supposed to be converted into hex and not the other way around. I even tried to enter the hex instead of passwd, still no success.  And then 'could not open keymaster device in keystore (Operation not permitted). 

By the way, I think those blobs in root directory are not needed as long as you simlink them (they could be taken from /system/vendor/firmware.

**Comment by nailyk-fr** [comment]
@optimumpr 

> By the way, I think those blobs in root directory are not needed as long as you simlink them (they could be taken from /system/vendor/firmware.

The only symlink I made is in /system (bin, lib, vendor & etc). It is a workaround for blobs availability, even if /system is not mounted and do not prevent /system being mounted. You found a symlink elsewhere? 
Have you try decrypt after mounting /system (this should use firmware blobs of your rom) ? Maybe without adding it in recovery ramdisk. 
Can you send me your build? Maybe I can check some points after ramdisk extraction. I also should be able to launch your twrp binary as we have same soc. 
Can you try to make a build for my z3? Perhaps that can point a problem/solution? 

@Micha-Btz 

> why you don't fork the sony rhine repo

@optimumpr 

> I got your github notification. 

With lot of trouble it is done. I think you should push code modifications. Sorry if you receive a lot of notifications but I deal with a lot of `! [rejected]        cm-12.1 -> cm-12.1 (non-fast-forward)` :scream_cat:

**Comment by nailyk-fr** [comment]
@optimumpr 
did you launch the build with `mka recovery-full` and see the [`done`](https://github.com/nailyk-fr/android_device_sony_rhine-common/blob/cm-12.1/recovery/Android.mk#L251) at the end of the build?

**Comment by optimumpr** [comment]
@nailyk-fr 
I do make -j# recoveryimage and it builts fine. I also removed recovery related executables and static libraries in objects to get a clean build. 

I'll try to build for Z3...

**Comment by nailyk-fr** [comment]
@optimumpr 
So if you doesn't use the `recovery-full` I made in the recovery/Android.mk (lines [55](https://github.com/nailyk-fr/android_device_sony_rhine-common/blob/cm-12.1/recovery/Android.mk#L55) to 79) you need to: 
build keystore / keystore.default / keystore.msm8974 , then cp all files to your ramdisk in addition of all the blobs, then `mka recoveryimage` because hardware keymaster is open by the keystore 'application'. 
Think that is your 'permission error': file not found. You can check presence of /bin/keystore binary in your ramdisk to verify the whole process.

**Comment by optimumpr** [comment]
@nailyk-fr 

Ok. I did a full build and although there is still no decryption, at least permission denied for keymaster is gone. Now the log says signing with keymaster, but signing fails. I think the reason is that I recently modified keymaster to have 256 bit decryption instead of 128.  The rom I use has 128 bit keymaster, but the blobs in twrp are 256. So, I'll have to do a few things and tomorrow we'll see what happens.  Thank you.

**Comment by optimumpr** [comment]
Success!!!   I did replace the keymaster blobs with those from my specific rom and now I have decryption at the first interface in TWRP without pressing cancel and manually mounting System. Thanks again.

**Comment by nailyk-fr** [comment]
@optimumpr 

> Success!!!

Nice work! I'm glad it work for you too! I'm sorry it take all this time. 

> Thanks again.

No problem, it is always pleasant to help and contribute. 

Now we have z1/z1c/z3 working, is it conceivable we try to made 'clean patch' and push it to the official twrp branch? (z3 is actually not officially supported)
Maybe after we can try to extend it to other hardware-encrypted qualcom phones?

**Comment by optimumpr** [comment]
Yes. Z1 is just recently got supported, but no one works on encryption there. And I am not sure they would accept the patch, because they think that nobody uses encryption anyway. By the way, our phones are not exactly hardware-encrypted: it is just signing that is done through a certificate sitting in Qualcom module. You noticed that in a log it says 'has_hw_crypto 0'.  For this to be active, you need changes in kernel configuration and trees to activate dm_req developed by Qualcom.

**Comment by optimumpr** [comment]
Also, if you know, now that we can mount data in TWRP:  you could potentially install another rom without re-encrypting the phone. Have you tried that? TWRP, when formatting data, wipes encryption, which is not good. We would have to revert that commit https://gerrit.omnirom.org/#/c/14353/1/partition.cpp . I actually can't imagine why they would need it in the first place? On kitkat, I had the phone encrypted once only and then could change roms without wiping cryptfooter.

**Comment by nailyk-fr** [comment]
@optimumpr 

> I am not sure they would accept the patch

:worried: I think encryption is more and more used. Some manufacturers are also factory encrypting. I think it should be functional in the high-quality recovery twrp is. But it is only my opinion. 

> By the way, our phones are not exactly hardware-encrypted: it is just signing that is done through a certificate sitting in Qualcom module. You noticed that in a log it says 'has_hw_crypto 0'. For this to be active, you need changes in kernel configuration and trees to activate dm_req developed by Qualcom.

You are totally right and it certainly will be my next work. The hardware random generator should be another one. [Qcom ad](https://www.qualcomm.com/products/snapdragon/security)

> when formatting data, wipes encryption, which is not good

I think it is a good solution because that is the only way to decrypt a phone. But I suppose, if you want keep encryption and format data, you can use the '[Change filesystem](http://i.imgur.com/ATzoYVa.png)' button (not tested!). It should only wipe your data and not the footer. Another solution is to go in your settings and select '[use rm -rf instead of formatting](http://i.imgur.com/aWVcTGF.jpg)' then click on wipe. 

edit: link 'change filesystem' replaced

**Comment by nailyk-fr** [comment]
in this [comment](https://github.com/TeamWin/android_device_oneplus_oneplus2/issues/1#issuecomment-171717595) @jfeise had same problem as I had: copying files when building with 'mka recoveryimage'. 
For him, PRODUCT_COPY_FILES seems to be working but not for me. Also as I need to create symlink, folders etc, is there a way to work with prebuild.mk? ([reference](http://stackoverflow.com/questions/14687287/error-adding-prebuilt-apk-with-shared-libraries-to-aosp/14691504#14691504))

**Comment by Micha-Btz** [comment]
yesterday i tried to build this from scratch, sadly the build doesn't run fine. for the moment i'm not able to build a cm12.1. so more tests are not possible atm.
https://gist.github.com/Micha-Btz/3acfad58e2bad25ba592

**Comment by Micha-Btz** [comment]
@optimumpr 

> If you don't have a pin setup before encrypting, then your data partition is encrypted with default >password, which is literally 'default_password'. Depending on your rom, you might have settings (in >security) to have the system ask for default password, but that is no security at all. You should do >encryption over adb to have one long password for boot/encryption and another short one to unlock >screen. You can head for XDA Security discussion thread where I have a detailed guide on that. 

can you please point me to the right thread?

**Comment by nailyk-fr** [comment]
@Micha-Btz 
I just give a try and everything goes fine: 
1. [Manifests](http://paste.debian.net/365986/) (Widget should be for my maguro as I remember but it is still in manifest)
2. `repo sync`
2. `breakfast amami`
3. `mka -j5 recovery-full`
4. and the [result](http://paste.debian.net/365987/)
So I doesn't understand why you have a error with libta. 
This kind of errors often appears when my $OUT folder is not clean after a `repo sync`. I think you could do `rm -r $OUT` (warning if $OUT is not in your env!); should be replaced with `rm -r /home/mdomann/handy/build/cm12.1/out/target/product/amami` in your case. 
Hope this help.

**Comment by optimumpr** [comment]
@Micha-Btz 

Here is the XDA link:  http://forum.xda-developers.com/general/security/guide-separate-passwords-encrypted-t3048072

**Comment by nailyk-fr** [comment]
Do you think I can try the workaround with this [issue](https://github.com/TeamWin/Team-Win-Recovery-Project/issues/484#issue-104257049) ? (The s4 is also [msm8974](http://www.gsmarena.com/samsung_i9506_galaxy_s4-5542.php) based)

**Comment by Micha-Btz** [comment]
@nailyk-fr 
recovery build for me doesn't work
the twrp.fstab -> ../../honami/recovery/root/etc/twrp.fstab is linked to honami device.
can you please fork the amami repo too and commit your patch there. or post the complete patch. the link above is only to the patch head.

https://gist.github.com/Micha-Btz/b1724d0ac736bf48ce3e

**Comment by nailyk-fr** [comment]
I push [that](https://github.com/nailyk-fr/android_device_sony_rhine-common/commit/a6748f31f18f9f193bc53d382714d427c8e98c2f). Sorry. Is this ok?

**Comment by Micha-Btz** [comment]
build run successfully, but im not sure if it will work.
there are some errors while kernel generation.
https://gist.github.com/Micha-Btz/9703965351f7bbadb9b9

tested, doesn't boot. only the scrren get on, thats it.

**Comment by optimumpr** [comment]
@Micha-Btz 

Your kernel errors are no problem.  Now, CM based kernels have no modules, so this is disabled. Regarding booting.  If you have a kernel with builtin recovery and also recovery in Fota/Recovery partition, you will have no boot in recovery. So, you either need to remove recovery from boot image or remove recovery in Fota. To do the latter, go to fastboot and execute fastboot erase recovery.  Then your built in (in boot image) recovery will work.

**Comment by nailyk-fr** [comment]
@optimumpr 
With the new sony bootloader, boot in recovery is provided.

@Micha-Btz 
I also have this kernel modules disabled 'error', but everything goes fines and you report it working. If something goes wrong, turn off your phone and reflash a know working recovery (on z3 I have a hardware power off button near the sdcard slot, maybe you have something similar?)

**Comment by nailyk-fr** [comment]
@optimumpr 
I see your [rom thread](http://forum.xda-developers.com/xperia-z1/orig-development/rom-aosp-jaguar-lp-5-1-1-r13-official-t3188925) (thank for the quote) but can't find your source code. How do you automated file copying in your twrp building? 

After reading 'build/core/Makefile' I'm still stuck. PRODCUT_COPY_FILES only copy when full build is done. `recovery_root_private` / `recovery_ressource_private` had no effect. 
I also read `device/sony/common/extract_elf_ramdisk/Android.mk`, it is build on recovery generation because it specify `LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES`. 
I try to switch my `twrp.fstab` from [`ETC`](https://github.com/nailyk-fr/android_device_sony_shinano-common/blob/cm-12.1/recovery/Android.mk#L15) to `RECOVERY_EXECUTABLES` (with no hope) and it doesn't work. 
- Can anyone explain to me how to add build dependencies to `recoveryimage` when I want to modify files in my device tree only?

**Comment by Micha-Btz** [comment]
> Install: /home/mdomann/handy/build/cyano12.1/out/target/product/amami/system/lib/libcryptfslollipop.so
> cp: der Aufruf von stat für „/home/mdomann/handy/build/cyano12.1/out/target/product/amami/recovery/root/etc/bootrec-device“ ist nicht möglich: Datei oder Verzeichnis nicht gefunden
> device/sony/msm8974-common/boot/custombootimg.mk:17: recipe for target '/home/mdomann/handy/build/cyano12.1/out/target/product/amami/boot.img' failed
> make: **\* [/home/mdomann/handy/build/cyano12.1/out/target/product/amami/boot.img] Error 1
> make: **\* Warte auf noch nicht beendete Prozesse...

every time at first compile attempt, i get the error above. i think there is the path wrong.
i know this is also wip but i think there shouldn't be path to one device. since hine-common is used for many devices.
https://github.com/nailyk-fr/android_device_sony_rhine-common/commit/1b3df91c4398f7261e49ebec97588fa222fe7819
i think this should copy device/sony/amami/recovery/root/etc/bootrec-device like in your commit, but it didn't.

**Comment by nailyk-fr** [comment]
@Micha-Btz 

> since hine-common is used for many devices.

you are totally right! I think there is two way: 
1. replace by device/sony/${CM_BUILD}/recovery/root/etc/bootrec-device
2. link original file in the rhine folder, then cp the linked filed. 

The first one seems cleaner (I'm currently testing it)
Do you think there is another way to do it? 

> i think this should copy device/sony/amami/recovery/root/etc/bootrec-device like in your commit, but it didn't.

Did you launch build with `mka recovery-full` ? In theory recovery-full should build twrp.fstab, then recoveryimage. 

Thank for your feedback!
P.S.: @Micha-Btz you have write access to the repo so you can push commit

**Comment by nailyk-fr** [comment]
It work for me now. Here is the [commit](https://github.com/nailyk-fr/android_device_sony_rhine-common/commit/573247924862077cfce74b8c371b8aeafd8a1cf3)

**Comment by Dees-Troy** [comment]
CM decided to stop supporting PRODUCT_COPY_FILES for recovery builds. Instead the recommended approach for files needing to be in the recovery ramdisk is to place the files in a recovery/root subfolder which will be copied into the ramdisk near the end of the recoveryimage build. You can see an example of that approach in this commit for the OnePlus One:
https://github.com/TeamWin/android_device_oneplus_bacon/commit/e195705c72148725336e2a7494cb7024b3d17939

I fixed decrypt on the OnePlus One in CM 12.1 a long time ago (April or May of 2015). The recent push there to the bacon device configs should make decrypt work again with the latest CM 13 nightlies.

The vast majority of devices do not use Qualcomm's hardware backed encryption that is enabled with the TARGET_HW_DISK_ENCRYPTION := true. I only know of 2 "devices" using it at this point: the OnePlus One and the special hammerheadcaf version of CM for the Nexus 5. Google tried to get the hardware encryption to work on shamu Nexus 6 but they were unable to solve a deep sleep issue on that device.

There are quite a few devices that use the hardware backed keymaster now with 5.0 and higher. Each device requires a certain amount of work to figure out which proprietary files are needed to make decrypt work in TWRP. We have about 3 developers total on our team, none of which are full-time. We work on TWRP as a hobby because we like to do it. We also give away our software free of charge. We neither have the money to buy all of the devices or the time to work with remote testers to make decrypt work on all of them. If you like working on this stuff and you get it working on a device, feel free to submit your changes to http://gerrit.twrp.me to be reviewed and merged into our regular device repos.

You are welcome to continue using this issue as a "forum" if you like, but I am closing this particular issue because it was solved a long time ago.

**Comment by nailyk-fr** [comment]
@Dees-Troy 
Thank a lot for your help! 
I don't know that github issue can't be used this way. 

@Diapolo
I had recently open this xda [thread](http://forum.xda-developers.com/z3/development/z3-twrp-2-8-7-0-d6603-t3273996). If you like, we can continue investigations/discuss on it. 

Really sorry for inconvenience.
